src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.cs

```diff
@@ -221,47 +297,80 @@ internal void ClearAnalyzerState(ImmutableArray<DiagnosticAnalyzer> analyzers)
         {
             if (!analyzers.IsDefaultOrEmpty)
             {
-                foreach (var analyzer in analyzers)
-                {
-                    ClearDescriptorState(analyzer);
-                    ClearAnalysisScopeState(analyzer);
-                }
+                ClearDescriptorState(analyzers);
+                ClearAnalysisScopeState(analyzers);
             }
         }
 
-        private void ClearDescriptorState(DiagnosticAnalyzer analyzer)
+        private void ClearDescriptorState(ImmutableArray<DiagnosticAnalyzer> analyzers)
         {
-            // Host is disposing the analyzer instance, unsubscribe analyzer exception handlers.
-            Tuple<ImmutableArray<DiagnosticDescriptor>, EventHandler<Exception>> value;
-            if (_descriptorCache.TryRemove(analyzer, out value))
+            lock (_descriptorMap)
             {
-                var descriptors = value.Item1;
-                var handler = value.Item2;
-                if (handler != null)
+                foreach (var analyzer in analyzers)
                 {
-                    foreach (var descriptor in descriptors)
+                    // Host is disposing the analyzer instance, unsubscribe analyzer exception handlers.
+                    Tuple<ImmutableArray<DiagnosticDescriptor>, EventHandler<Exception>> value;
+                    if (_descriptorMap.TryGetValue(analyzer, out value))
                     {
-                        descriptor.Title.OnException -= handler;
-                        descriptor.MessageFormat.OnException -= handler;
-                        descriptor.Description.OnException -= handler;
+                        var descriptors = value.Item1;
+                        var handler = value.Item2;
+                        if (handler != null)
+                        {
+                            foreach (var descriptor in descriptors)
+                            {
+                                descriptor.Title.OnException -= handler;
+                                descriptor.MessageFormat.OnException -= handler;
+                                descriptor.Description.OnException -= handler;
```

Comment:
Yes, arguments to the analyzer's supported descriptors are of `LocalizableString` type, which is a public abstract type. We do provide an exception safe [default implementation](http://source.roslyn.io/#Microsoft.CodeAnalysis/Diagnostic/LocalizableResourceString.cs,5febf21f2f3ec86c,references) based on resource strings, but 3rd parties can provide their own implementation which can throw. We need to catch those exceptions and log them as analyzer exception diagnostics.

Unit tests for verifying this scenario: http://source.roslyn.io/#Roslyn.Compilers.UnitTests/Diagnostics/DiagnosticLocalizationTests.cs,12bc2060edf8e15d